### ISSUE TITLE
How to load and use LiDAR PCD's with intensity for registration?

### ISSUE THREAD

**Issue by saimanoj18** [IntelVCL/Open3D#125]
Hi,
Thank you for this great library!

I would like to use lidar point clouds from velodyne sensor with Open3D for point cloud registration. These lidar point clouds come with intensity as well. 

It seems like PointCloud class accounts for points, normals and color only. Am I correct ? 

If so, do you have any suggestion on how can I **load and use** use intensity information for point cloud registration ? Simple thought that comes to mind is just equate the RGB values to the intensity values. Do you have any other suggestions ?

A sample cloud can be downloaded from here.
https://drive.google.com/file/d/1UJ19SU6oYYui7SI5nRkRkfcE32lcR_SE/view?usp=sharing

I am experimenting by copying the values of intensity to RGB fields, then what would be right thing to do ? Should I copy equally to all RGB fields or should I just copy to one field and leave others as zero ? 

The lidar intensity values range from 0 to 255 based on the reflectivity/intensity. Do I need to scale them ? 

Finally, my application is to employ colored point cloud registration from Open3D library, hence what might be the right way of using the intensity ?

https://drive.google.com/file/d/1fmeBB6JaW-Ia0T9UEMbYR9eF7_8Yf9sf/view?usp=sharing
A pair of point clouds that I am trying to register can be downloaded from the above link. Specifically, I have stored a PCD file with XYZ and intensity as XYZRGB where RGB values are equal to intensity.

After running the colored_point_cloud_registration on this pair, point to plane based ICP seems to be converging, but colored point cloud registration completely misaligns the point clouds. The possible reason can be that the intensity values might be noisy from the lidar sensor, but is there any way to strike balance between geometric and color minimization terms ?

Do you think, you can tweak parameters from your experience to register these pair of point clouds ?

Laslty, in the below two lines (taken from colored point cloud registration tutorial)

< print("   clouds to refine the alignment. Distance threshold 0.02.")
  result_icp = registration_icp(source, target, 0.02, current_transformation, TransformationEstimationPointToPlane())>

The parameter 0.02 is the maximum distance that is used to find correspondences ? I have a small doubt in this. Is this 0.02, the max euclidean distance between two points (x,y,z), to be assigned as correspondences ? Do you use normal information in finding correspondences in PointToPlace metric ? In that case, how is this 0.02 being calculated or used to find correspondences ?  

Thank you so much for your time!

Thank you,
Manoj

**Comment by qianyizh** [comment]
If you use point cloud with intensity, just copy the intensity value to all the rgb fields (``PointCloud.colors``). We do require that the values to be scaled to 0~1.

I have checked your point clouds to be registered. They do not have normal. In addition, they are sparse, so I don't think the normal estimation will work for them. In this case, your best bet is the point_to_point icp. Both the point_to_plane icp and colored point cloud registration requires normal information. Thus I don't think they fit in this use case.

The distance threshold is the absolute Euclidean distance. We don't use normal to determin correspondences. Normal is used only in optimization. That's also part of the reason why point_to_point icp and point_to_plane icp share the same function, just a different parameter (which is used to compute the gradient of the optimization objective).

**Comment by syncle** [comment]
Yes. I think the point cloud you shared is too sparse. It looks even sparser than ordinal Velodyne scan. Did you subsample Lidar scans? 

To fully utilize point-to-plane or colored point cloud registration, it is recommended to have dense point clouds. Once you have it, compute normal and visually inspect whether the estimated normal is reasonable [like this example](http://open3d.org/docs/tutorial/Basic/pointcloud.html#vertex-normal-estimation).

**Comment by saimanoj18** [comment]
Thank you both for those useful suggestions. Yes, point to point ICP worked reasonably. What I did was, I tried to create a rough alignment or a rough submap, estimate normals from that and then associate them to scan and use point to plane ICP. Then it worked. Thanks for explaining about distance threshold. Yes, I did some cropping of lidar scans and the lidar is placed in horizontal position and hence the number of scans are much lower. I could get a reasonable alignment but not perfect in terms of road markings.
